src/components/implementation/no_interface/scheddev/sl.c

```diff
@@ -248,18 +249,24 @@ sl_sched_loop(void)
 			 * states of it's child threads) and normal notifications (mainly activations from
 			 * it's parent scheduler).
 			 */
-			pending = cos_sched_rcv(BOOT_CAPTBL_SELF_INITRCV_BASE, &tid, &blocked, &cycles);
+			pending = cos_sched_rcv(BOOT_CAPTBL_SELF_INITRCV_BASE, RCV_ALL_PENDING,
+						&rcvd, &tid, &blocked, &cycles);
 			if (!tid) continue;
 
 			t = sl_thd_lkup(tid);
 			assert(t);
 			/* don't report the idle thread */
 			if (unlikely(t == sl__globals()->idle_thd)) continue;
+
+			if ((ret = sl_cs_enter_sched()) == -EBUSY) continue;
 			sl_mod_execution(sl_mod_thd_policy_get(t), cycles);
 			if (blocked) sl_mod_block(sl_mod_thd_policy_get(t));
 			else         sl_mod_wakeup(sl_mod_thd_policy_get(t));
+
+			sl_cs_exit();
 		} while (pending);
 
+		if ((ret = sl_cs_enter_sched()) == -EBUSY) continue;
 		/* If switch returns an inconsistency, we retry anyway */
 		sl_cs_exit_schedule_nospin();
 	}
```

Comment:
The usage of the `ret` variable in this function makes my head spin. Combined with the use of equals in if statements, it's really confusing. Is there any refactoring that could be done to make it more clear what's going on? The variable could at least be renamed...